Sprint 5/step 1

piladik opened the pull request.

piladik committed:
Redux Toolkit Migration 1.0
Переписал action'ы и reducer'ы для хранилища ingredients с использованием Redux Toolkit

piladik committed:
Redux Toolkit Migration 1.1
Переписал action'ы и reducer'ы для хранилища constructorBurger с использованием Redux Toolkit

piladik committed:
reduxToolkitMigration 2.0
Перенес весь код хранилищя на Redux Toolkit

piladik committed:
reduxToolkitMigration 3.0
1. Перенес хранилище на ts
2. Типизировал хуки useSelect, useDispatch кастомными хуками useAppSelect, useAppDispatch и заменил ими старые хуки
3. Удалил неиспользуемые интерфейсы и типы и добавил новые

piladik committed:
reduxToolkitMigration 3.1
Small fixes:
- Исправлен файл helpers.ts

piladik committed:
Верстка страницы /feed
В компонент /pages/feed-page.tsx импортируется два компонента FeedOrders и FeedStatus.
Первый отвечает за отображение заказов, второй за отображние готовности заказов.
Компоненты FeedOrders и FeedStatus также разделены на более мелкие компоненты.
FeedOrders импортирует компонент FeedOrder. Сейчас компонент работает с тестовыми данными.
В дальнейшем, когда мы будем получать данные с сервера, это упростит работу.
FeedStatus импортирует три компонента:
1. <OrdersStatus /> отображает готовые и готовящиеся заказы.
2. <OrdersFinishedAllTime /> отображает количство выполненных заказов за все время.
3. <OrdersFinishedToday /> отображает количство выполненых заказов за сегодня.

piladik committed:
Добавил скролбар к компоненту FeedOrders.
FeedOrder теперь создается при помощи Array.map, а не в ручную.
Добавил типизацию пропсов для FeedOrder.

piladik committed:
Верстка /feed/:id

piladik committed:
Исправление маленьких багов по верстке

piladik committed:
Верстка /profile/orders & /profile/orders/:id

piladik committed:
Доработка роутинга в приложении
При переходе на /feed/:id или /profile/order/:id по ссылке открывается отдельная страница, при клике открывается модальное окно

piladik committed:
Функциональность экрана «Лента заказов»
1. Написан socketMiddleware для работы с websocket
2. Написаны action'ы и reduser для ленты заказов

piladik committed:
Доработка ленты заказов
1. Написал функцию для расчета стоимости заказа по списку айди ингредиентов.
2 Написал функцию, которая в качестве аргументов принимает айди ингредиентов и список ингредиентов,
и возвращает объект, в котором содержится 6 url адресов картинок и количество картинок ингредиентов, которые не будут отображены.
3. Воспользовался компонентом из библиотеки яндекса для отображения даты создания заказа.
4. Добавил логику отрисовки количества неотображенных картинок.

piladik committed:
Полностью закончил работу над версткой и функционалом /feed & /feed/:id

piladik committed:
Добавлено динамическое отображение статуса заказа.

piladik committed:
Функциональность экрана «История заказов»
1. Написаны action'ы и reduser для истории заказов заказов
2. Middleware теперь обрабатывает два сокет соединения в зависимости от передавыемых экшенов
3. Поправлена верстка для Истории заказов - теперь заказы отображаются с ответа сервера.

andryxxa93 commented on src/services/middleware/socket-middleware.ts:
```diff
@@ -0,0 +1,74 @@
+          const { success, ...restParsedData } = parsedData;
+
+          if (success) {
+            dispatch(onMessage({ ...restParsedData, dataReceived: true }));
```
**Можно лучше**
Необходимо добавить проверку при получении сообщения от сокета. Eсли токен протух и приходит поле message 'Invalid or missing token' то его необходимо обновить через refreshToken.

andryxxa93 commented on src/services/slices/constructor.ts:
```diff
@@ -0,0 +1,40 @@
+      state.ingredients.bun = action.payload;
+    },
+    addFilling(state, action) {
+      state.ingredients.fillings.push({ ...action.payload, nanoid: nanoid() });
```
**Надо исправить**

редьюсер должен быть чистой функции, сейчас при каждом вызове добавления ингредиента результат будет отличаться, даже если стейт и экшен одинаковые - будет генерироваться каждый раз новый id,  при добавлении ингредиента в конструктор id нужно генерировать и передавать в экшене, а не генерировать в редьюсере. Иначе это делает редьюсер не чистой функцией, что нарушает правило редьюсеров

andryxxa93 reviewed (changes requested):
Здравствуйте! Проделана большая работа 👍 

Что понравилось:
- Раздельное хранение данных полеченных по сокету
- Хорошая типизация хранилища и хуков useDispatch и useSelector
- Сокеты открываются и закрываются согласно тех заданию

Однако есть пару моментов, которые необходимо исправить, чтобы работа была принята.

Необходимо исправить:
- При первом заходе в приложение у вас каждый запрос вызывается 4 раза. Всё из-за раздутого useEffect в `App`. Рекомендую его разбить на несколько небольших и чтобы у каждого был свой список зависимостей.
 
<img width="593" alt="Снимок экрана 2023-05-17 в 16 46 33" src="https://github.com/piladik/react-burger/assets/60173362/94874628-4fed-43eb-8917-09269d382689">

- При открытии конкретного заказа из Ленты заказов по прямой ссылке - пустая страница.

<img width="1026" alt="Снимок экрана 2023-05-17 в 16 36 18" src="https://github.com/piladik/react-burger/assets/60173362/39795eda-5420-44f4-b9da-9abc297b790d">

- при добавлении ингредиента в конструктор id нужно генерировать и передавать в экшене

Более подробно описание проблемы смотрите комментарии в pull request.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

**Напомню, чтобы работа была принята нужно исправить все замечания помеченные как **необходимо исправить**.

Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12) перед отправкой на ревью. 

Удачных правок!

piladik committed:
Поправки после первого ревью
1. Сокращено кол-во отправлеямых запросов при первом запуске приложения
2. Исправлена ошибка из-за которой открытие заказа /feed/:id по прямой ссылке не работало
3. constructorSlice теперь чистый редьюсер

piladik committed:
Добавил обновление токена в socketMiddleware при ответе с сервера "Invalid or missing token"

andryxxa93 reviewed (approved):
Супер, вы отлично поработали!
Поздравляю! Работа принята.
Желаю удачного продолжения обучения!
